core/executor/src/wasmi_runtimes_cache.rs

```diff
@@ -17,13 +17,14 @@
 //! Implements a cache for pre-created Wasm runtime module instances.
 
 use crate::error::Error;
-use crate::wasm_executor::WasmExecutor;
+use crate::native_executor::SubstrateRuntime;
```

Comment:
Can we not have a cache that works for every possible wasm executor?

Reply:
No, the cache is rather different between at least Wasmi and Wasmtime. In particular, the structs that are cached are completed different.

It's probably possible to make another trait for a SubstrateRuntime cache which returns different dynamically boxed SubstrateRuntimes depending on the implementation, but I haven't seen much benefit in that.

Reply:
Yeah I know that the inner structs would be different, anything else would not make any sense at all.

But given that we can reuse the rest of the logic, I don't see why we should split the stuff. You already have your `SubstrateRuntime` that basically abstracts over the cached runtime. You should probably just rename `SubstrateRuntime` to `CachedRuntime` and use this trait to store the runtimes internally.

Reply:
For the wasmtime runtime (and the WasmiRuntime too, but to a lesser degree), you can't just return the cached object. You use the cached object to construct the actual runtime. In the case of `WasmiRuntime` it has a reference to an executor, which is not cached. This is avoidable, but not for wasmtime.

Reply:
OK, I made the RuntimesCache store all types of runtimes.